File: crypto-trading-bot/src/connectors/exchanges/BaseExchange.js
const EventEmitter = require('events');
const ccxt = require('ccxt');

class BaseExchange extends EventEmitter {
    constructor(config) {
        super();
        this.config = config;
        this.exchange = null;
        this.isConnected = false;
        this.lastPriceUpdate = new Map();
        this.orderBook = new Map();
        this.balances = {};
        this.rateLimits = new Map();
        this.retryCount = 0;
        this.maxRetries = 3;
    }

    async initialize() {
        this.exchange = new ccxt[this.config.exchangeId]({
            apiKey: this.config.apiKey,
            secret: this.config.secret,
            password: this.config.passphrase,
            sandbox: false,
            enableRateLimit: true,
            timeout: 30000,
            options: {
                adjustForTimeDifference: true,
                recvWindow: 10000
            }
        });

        await this.exchange.loadMarkets();
        this.isConnected = true;
        await this.updateBalances();
        this.emit('connected');
    }

    async fetchTicker(symbol) {
        try {
            const ticker = await this.exchange.fetchTicker(symbol);
            this.lastPriceUpdate.set(symbol, Date.now());
            return ticker;
        } catch (error) {
            throw new Error(`Failed to fetch ticker for ${symbol}: ${error.message}`);
        }
    }

    async fetchOrderBook(symbol, limit = 20) {
        try {
            const orderBook = await this.exchange.fetchOrderBook(symbol, limit);
            this.orderBook.set(symbol, {
                ...orderBook,
                timestamp: Date.now()
            });
            return orderBook;
        } catch (error) {
            throw new Error(`Failed to fetch order book for ${symbol}: ${error.message}`);
        }
    }

    async createOrder(symbol, type, side, amount, price, params = {}) {
        try {
            const order = await this.exchange.createOrder(symbol, type, side, amount, price, params);
            this.emit('orderCreated', order);
            return order;
        } catch (error) {
            throw new Error(`Failed to create order: ${error.message}`);
        }
    }

    async createMarketBuyOrder(symbol, amount, price = null, params = {}) {
        return await this.createOrder(symbol, 'market', 'buy', amount, price, params);
    }

    async createMarketSellOrder(symbol, amount, price = null, params = {}) {
        return await this.createOrder(symbol, 'market', 'sell', amount, price, params);
    }

    async createLimitBuyOrder(symbol, amount, price, params = {}) {
        return await this.createOrder(symbol, 'limit', 'buy', amount, price, params);
    }

    async createLimitSellOrder(symbol, amount, price, params = {}) {
        return await this.createOrder(symbol, 'limit', 'sell', amount, price, params);
    }

    async cancelOrder(id, symbol) {
        try {
            const result = await this.exchange.cancelOrder(id, symbol);
            this.emit('orderCancelled', { id, symbol });
            return result;
        } catch (error) {
            throw new Error(`Failed to cancel order ${id}: ${error.message}`);
        }
    }

    async fetchOrder(id, symbol) {
        try {
            return await this.exchange.fetchOrder(id, symbol);
        } catch (error) {
            throw new Error(`Failed to fetch order ${id}: ${error.message}`);
        }
    }

    async fetchOpenOrders(symbol = null) {
        try {
            return await this.exchange.fetchOpenOrders(symbol);
        } catch (error) {
            throw new Error(`Failed to fetch open orders: ${error.message}`);
        }
    }

    async fetchBalance() {
        try {
            this.balances = await this.exchange.fetchBalance();
            return this.balances;
        } catch (error) {
            throw new Error(`Failed to fetch balance: ${error.message}`);
        }
    }

    async updateBalances() {
        await this.fetchBalance();
        this.emit('balanceUpdated', this.balances);
    }

    getBalance(currency) {
        return this.balances[currency] || { free: 0, used: 0, total: 0 };
    }

    async getPrice(symbol) {
        const cached = this.lastPriceUpdate.get(symbol);
        if (cached && Date.now() - cached < 5000) {
            const orderBook = this.orderBook.get(symbol);
            if (orderBook && orderBook.bids.length > 0 && orderBook.asks.length > 0) {
                return (orderBook.bids[0][0] + orderBook.asks[0][0]) / 2;
            }
        }

        const ticker = await this.fetchTicker(symbol);
        return ticker.last;
    }

    async getLiquidity(symbol) {
        const orderBook = await this.fetchOrderBook(symbol, 10);
        const bidLiquidity = orderBook.bids.reduce((sum, [price, amount]) => sum + amount, 0);
        const askLiquidity = orderBook.asks.reduce((sum, [price, amount]) => sum + amount, 0);
        return Math.min(bidLiquidity, askLiquidity);
    }

    async getSpread(symbol) {
        const orderBook = await this.fetchOrderBook(symbol, 1);
        if (orderBook.bids.length === 0 || orderBook.asks.length === 0) {
            return 0;
        }
        const bid = orderBook.bids[0][0];
        const ask = orderBook.asks[0][0];
        return (ask - bid) / bid;
    }

    async withdraw(currency, amount, address, tag = null, params = {}) {
        try {
            const withdrawal = await this.exchange.withdraw(currency, amount, address, tag, params);
            this.emit('withdrawalCreated', withdrawal);
            return withdrawal;
        } catch (error) {
            throw new Error(`Failed to withdraw ${amount} ${currency}: ${error.message}`);
        }
    }

    async fetchDeposits(currency = null, since = null, limit = null, params = {}) {
        try {
            return await this.exchange.fetchDeposits(currency, since, limit, params);
        } catch (error) {
            throw new Error(`Failed to fetch deposits: ${error.message}`);
        }
    }

    async fetchWithdrawals(currency = null, since = null, limit = null, params = {}) {
        try {
            return await this.exchange.fetchWithdrawals(currency, since, limit, params);
        } catch (error) {
            throw new Error(`Failed to fetch withdrawals: ${error.message}`);
        }
    }

    async retryOperation(operation, ...args) {
        for (let i = 0; i < this.maxRetries; i++) {
            try {
                return await operation.apply(this, args);
            } catch (error) {
                if (i === this.maxRetries - 1) throw error;
                await this.sleep(Math.pow(2, i) * 1000);
            }
        }
    }

    checkRateLimit(endpoint) {
        const lastCall = this.rateLimits.get(endpoint);
        const minInterval = this.exchange.rateLimit || 1000;
        
        if (lastCall && Date.now() - lastCall < minInterval) {
            return false;
        }
        
        this.rateLimits.set(endpoint, Date.now());
        return true;
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    async executeArbitrage(buySymbol, sellSymbol, amount) {
        try {
            const buyOrder = await this.createMarketBuyOrder(buySymbol, amount);
            const sellOrder = await this.createMarketSellOrder(sellSymbol, amount);
            
            return {
                buyOrder,
                sellOrder,
                profit: sellOrder.average - buyOrder.average,
                timestamp: Date.now()
            };
        } catch (error) {
            throw new Error(`Arbitrage execution failed: ${error.message}`);
        }
    }

    async getMarketDepth(symbol, limit = 50) {
        const orderBook = await this.fetchOrderBook(symbol, limit);
        
        let bidDepth = 0;
        let askDepth = 0;
        
        for (const [price, amount] of orderBook.bids) {
            bidDepth += price * amount;
        }
        
        for (const [price, amount] of orderBook.asks) {
            askDepth += price * amount;
        }
        
        return { bidDepth, askDepth, total: bidDepth + askDepth };
    }

    async waitForOrderFill(orderId, symbol, timeout = 30000) {
        const startTime = Date.now();
        
        while (Date.now() - startTime < timeout) {
            const order = await this.fetchOrder(orderId, symbol);
            
            if (order.status === 'closed') {
                return order;
            }
            
            await this.sleep(1000);
        }
        
        throw new Error(`Order ${orderId} not filled within timeout`);
    }

    disconnect() {
        this.isConnected = false;
        this.emit('disconnected');
    }

    getExchangeInfo() {
        return {
            id: this.config.exchangeId,
            name: this.exchange.name,
            countries: this.exchange.countries,
            rateLimit: this.exchange.rateLimit,
            fees: this.exchange.fees,
            markets: Object.keys(this.exchange.markets)
        };
    }
}

module.exports = BaseExchange;

File: src/exchanges/base_exchange.py
import asyncio
from abc import ABC, abstractmethod
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from utils.logger import get_logger

logger = get_logger(__name__)

@dataclass
class OrderBook:
    bids: List[Tuple[float, float]]
    asks: List[Tuple[float, float]]
    timestamp: float

@dataclass
class Ticker:
    symbol: str
    bid: float
    ask: float
    last: float
    volume: float
    timestamp: float

@dataclass
class Order:
    id: str
    symbol: str
    side: str
    amount: float
    price: float
    status: str
    timestamp: float

class BaseExchange(ABC):
    def __init__(self, name: str):
        self.name = name
        self.connected = False
        self.rate_limiter = None
        self.fees = {}
        self.markets = {}

    @abstractmethod
    async def initialize(self):
        pass

    @abstractmethod
    async def get_ticker(self, symbol: str) -> Optional[Ticker]:
        pass

    @abstractmethod
    async def get_order_book(self, symbol: str, limit: int = 20) -> Optional[OrderBook]:
        pass

    @abstractmethod
    async def get_all_tickers(self) -> Dict[str, Ticker]:
        pass

    @abstractmethod
    async def place_order(self, symbol: str, side: str, amount: float, 
                         price: Optional[float] = None, order_type: str = "market") -> Order:
        pass

    @abstractmethod
    async def cancel_order(self, order_id: str, symbol: str) -> bool:
        pass

    @abstractmethod
    async def get_order_status(self, order_id: str, symbol: str) -> Optional[Order]:
        pass

    @abstractmethod
    async def get_balance(self, currency: Optional[str] = None) -> Dict[str, float]:
        pass

    async def place_market_order(self, symbol: str, side: str, amount: float) -> Order:
        return await self.place_order(symbol, side, amount, None, "market")

    async def place_limit_order(self, symbol: str, side: str, amount: float, price: float) -> Order:
        return await self.place_order(symbol, side, amount, price, "limit")

    def get_trading_fee(self, symbol: str, side: str) -> float:
        return self.fees.get(symbol, {}).get(side, 0.001)

    def calculate_fee(self, symbol: str, side: str, amount: float, price: float) -> float:
        fee_rate = self.get_trading_fee(symbol, side)
        return amount * price * fee_rate

    async def get_best_bid_ask(self, symbol: str) -> Optional[Tuple[float, float]]:
        ticker = await self.get_ticker(symbol)
        if ticker:
            return ticker.bid, ticker.ask
        return None

    async def get_mid_price(self, symbol: str) -> Optional[float]:
        bid_ask = await self.get_best_bid_ask(symbol)
        if bid_ask:
            return (bid_ask[0] + bid_ask[1]) / 2
        return None

    def is_connected(self) -> bool:
        return self.connected

    async def test_connection(self) -> bool:
        try:
            await self.get_balance()
            return True
        except:
            return False

    async def wait_for_rate_limit(self):
        if self.rate_limiter:
            await self.rate_limiter.wait()

    def _validate_symbol(self, symbol: str) -> bool:
        return symbol in self.markets

    def _validate_side(self, side: str) -> bool:
        return side.lower() in ['buy', 'sell']

    def _validate_amount(self, amount: float) -> bool:
        return amount > 0

    def _validate_price(self, price: Optional[float]) -> bool:
        return price is None or price > 0

    async def _handle_error(self, error: Exception, operation: str):
        logger.error(f"{self.name} {operation} error: {error}")
        
        if "rate limit" in str(error).lower():
            await asyncio.sleep(1)
        elif "insufficient" in str(error).lower():
            logger.warning(f"{self.name} insufficient balance")
        elif "invalid" in str(error).lower():
            logger.warning(f"{self.name} invalid request: {error}")

    async def shutdown(self):
        self.connected = False
        logger.info(f"{self.name} exchange connection closed")

    def __str__(self):
        return f"{self.name}Exchange(connected={self.connected})"